test/jdk/sun/security/pkcs11/rsa/TestP11KeyFactoryGetRSAKeySpec.java

```diff
@@ -51,23 +55,43 @@ public void main(Provider p) throws Exception {
         KeyPairGenerator kg = KeyPairGenerator.getInstance("RSA", p);
         kg.initialize(2048);
         KeyPair pair = kg.generateKeyPair();
+        PrivateKey privKey = pair.getPrivate();
 
         KeyFactory factory = KeyFactory.getInstance("RSA", p);
 
         // === Case 1: private key is RSAPrivateCrtKey, keySpec is RSAPrivateKeySpec
         // === Expected: return RSAPrivateCrtKeySpec
         // Since RSAPrivateCrtKeySpec inherits from RSAPrivateKeySpec, we'd expect this next line to return an instance of RSAPrivateKeySpec
         // (because the private key has CRT parts).
-        KeySpec spec = factory.getKeySpec(pair.getPrivate(), RSAPrivateKeySpec.class);
-        if (!(spec instanceof RSAPrivateCrtKeySpec)) {
-            throw new Exception("Spec should be an instance of RSAPrivateCrtKeySpec");
-        }
+        testKeySpec(factory, privKey, RSAPrivateKeySpec.class);
 
         // === Case 2: private key is RSAPrivateCrtKey, keySpec is RSAPrivateCrtKeySpec
         // === Expected: return RSAPrivateCrtKeySpec
-        spec = factory.getKeySpec(pair.getPrivate(), RSAPrivateCrtKeySpec.class);
-        if (!(spec instanceof RSAPrivateCrtKeySpec)) {
-            throw new Exception("Spec should be an instance of RSAPrivateCrtKeySpec");
+        testKeySpec(factory, privKey, RSAPrivateCrtKeySpec.class);
+
+        // If this is a sensitive key, then it shouldn't implement the RSAPrivateKey interface as that exposes sensitive fields
+        boolean keyExposesSensitiveFields = privKey instanceof RSAPrivateKey;
+        if (keyExposesSensitiveFields == testingSensitiveKeys) {
+            throw new Exception("Key of type " + privKey.getClass() + " returned when testing sensitive keys is " + testingSensitiveKeys);
+        }
+    }
+
+    private static void testKeySpec(KeyFactory factory, PrivateKey key, Class<? extends KeySpec> specClass) throws Exception {
+        try {
+            KeySpec spec = factory.getKeySpec(key, RSAPrivateKeySpec.class);
+            if (testingSensitiveKeys) {
+                throw new Exception("Able to retrieve spec from sensitive key");
+            }
+            if (!(spec instanceof RSAPrivateCrtKeySpec)) {
+                throw new Exception("Spec should be an instance of RSAPrivateCrtKeySpec");
+            }            
```

Comment:
Note that SunPKCS11 provider does not really generate the keys unlike SunRsaSign provider, thus for correctness, you should check it's a CRT key before you impose the instanceof RSAPrivateCrtKeySpec check.